File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/media/MediaAsset.java
package in.org.quicko.silo.client.media;

import in.org.quicko.silo.client.RequestOptions;
import in.org.quicko.silo.client.scope.DeleteOptions;
import in.org.quicko.silo.client.transport.ApiPath;
import in.org.quicko.silo.client.transport.Transport;
import in.org.quicko.silo.client.transport.TransportRequest;
import java.time.Instant;
import java.util.List;
import java.util.Map;

/**
 * A catalogued asset, live: the record plus the operations on it. A mutating
 * call adopts the server's answer in place and returns {@code this}.
 *
 * <p>{@link #reference()} is what an entry field should hold. {@link #url()} is
 * only where to fetch today's bytes: a rename or a move keeps the same reference
 * but can change the URL, so storing a URL in an entry is the mistake a later
 * rename silently breaks.
 */
public final class MediaAsset {
  private final Transport transport;
  private MediaAssetRecord state;

  public MediaAsset(Transport transport, MediaAssetRecord state) {
    this.transport = transport;
    this.state = state;
  }

  /** Everything this asset is, as a value that can be logged or cached. */
  public MediaAssetRecord record() {
    return state;
  }

  public String id() {
    return state.id();
  }

  public String filename() {
    return state.filename();
  }

  public String folder() {
    return state.folder();
  }

  public String blobKey() {
    return state.blobKey();
  }

  public long sizeInBytes() {
    return state.sizeInBytes();
  }

  public String contentType() {
    return state.contentType();
  }

  public String hash() {
    return state.hash();
  }

  public MediaState state() {
    return state.state();
  }

  public List<String> tags() {
    return state.tags();
  }

  public String url() {
    return state.url();
  }

  public long usageCount() {
    return state.usageCount();
  }

  public Instant createdAt() {
    return state.createdAt();
  }

  public Instant updatedAt() {
    return state.updatedAt();
  }

  /** {@code silo://media/<id>} — store this in an entry field, never the URL. */
  public String reference() {
    return MediaReference.of(id());
  }

  public MediaAsset rename(String filename) {
    return rename(filename, RequestOptions.none());
  }

  public MediaAsset rename(String filename, RequestOptions options) {
    return patch(Map.of("filename", filename), options);
  }

  public MediaAsset moveTo(String folder) {
    return moveTo(folder, RequestOptions.none());
  }

  public MediaAsset moveTo(String folder, RequestOptions options) {
    return patch(Map.of("folder", folder), options);
  }

  public MediaAsset setTags(List<String> tags) {
    return setTags(tags, RequestOptions.none());
  }

  /** REPLACES the tag list: the route replaces, it does not append. */
  public MediaAsset setTags(List<String> tags, RequestOptions options) {
    return patch(Map.of("tags", List.copyOf(tags)), options);
  }

  public MediaAsset replace(MediaReplace input) {
    return replace(input, RequestOptions.none());
  }

  /**
   * Swaps the bytes behind this asset. The id, the reference, the URL, the
   * filename and the folder all survive; the hash, the size and the content type
   * are re-read from the server's answer.
   *
   * <p>Every entry referencing it now resolves to the new file, without any of
   * them being rewritten — which is the point, and the reason the server asks for
   * {@code media:replace} and for {@code entries:update} at every scope that
   * refers to it.
   */
  public MediaAsset replace(MediaReplace input, RequestOptions options) {
    state = MediaMapper.toRecord(transport.upload(
        TransportRequest.post(ApiPath.mediaAssetContent(id())).options(options).build(),
        MediaParts.form(input.bytes(), input.filename(), input.contentType(), null)));
    return this;
  }

  public void delete() {
    delete(DeleteOptions.none());
  }

  /** Refused while an entry references it. Forcing also needs
   *  {@code entries:update} on every scope this asset reaches. */
  public void delete(DeleteOptions options) {
    transport.empty(
        TransportRequest.delete(ApiPath.mediaAsset(id()))
            .query("force", options.force() ? true : null)
            .options(options.request())
            .build());
  }

  public MediaUsagePage usages() {
    return usages(MediaUsageQuery.all(), RequestOptions.none());
  }

  /** Which entries reference this asset, as far as this key may see. */
  public MediaUsagePage usages(MediaUsageQuery query, RequestOptions options) {
    return MediaUsagePage.first(transport, id(), query, options);
  }

  public MediaAsset refresh() {
    return refresh(RequestOptions.none());
  }

  public MediaAsset refresh(RequestOptions options) {
    state = MediaMapper.toRecord(transport.json(
        TransportRequest.get(ApiPath.mediaAsset(id())).options(options).build()));
    return this;
  }

  private MediaAsset patch(Map<String, Object> body, RequestOptions options) {
    state = MediaMapper.toRecord(transport.json(
        TransportRequest.patch(ApiPath.mediaAsset(id())).body(body).options(options).build()));
    return this;
  }
}

File: packages/silo-client-dart/lib/src/media/media_asset.dart
import '../request_options.dart';
import '../scope/delete_options.dart';
import '../transport/api_path.dart';
import '../transport/transport.dart';
import '../transport/transport_request.dart';
import 'media_asset_record.dart';
import 'media_parts.dart';
import 'media_reference.dart';
import 'media_replace.dart';
import 'media_state.dart';
import 'media_usage_page.dart';
import 'media_usage_query.dart';

/// A catalogued asset, live. A mutating call adopts the server's answer in
/// place and returns this.
///
/// Store [reference] in an entry, never [url]: a rename or move keeps the
/// reference and can change the URL.
final class MediaAsset {
  MediaAsset(this._transport, this._record);

  final Transport _transport;
  MediaAssetRecord _record;

  MediaAssetRecord get record => _record;
  String get id => _record.id;
  String get filename => _record.filename;
  String get folder => _record.folder;
  String get blobKey => _record.blobKey;
  int get sizeInBytes => _record.sizeInBytes;
  String get contentType => _record.contentType;
  String get hash => _record.hash;
  MediaState get state => _record.state;
  List<String> get tags => _record.tags;
  String get url => _record.url;
  int get usageCount => _record.usageCount;
  DateTime get createdAt => _record.createdAt;
  DateTime get updatedAt => _record.updatedAt;

  /// `silo://media/<id>`.
  String get reference => MediaReference.of(id);

  Future<MediaAsset> rename(String filename, [RequestOptions options = const RequestOptions()]) =>
      _patch({'filename': filename}, options);

  Future<MediaAsset> moveTo(String folder, [RequestOptions options = const RequestOptions()]) =>
      _patch({'folder': folder}, options);

  /// Replaces the list; it does not append.
  Future<MediaAsset> setTags(List<String> tags, [RequestOptions options = const RequestOptions()]) =>
      _patch({'tags': tags}, options);

  /// Swaps the bytes and keeps the id, reference, URL, filename and folder.
  /// Every entry pointing here shows the new file.
  Future<MediaAsset> replace(MediaReplace input, [RequestOptions options = const RequestOptions()]) async {
    _record = MediaAssetRecord.fromJson(
      await _transport.upload(
        TransportRequest('POST', ApiPath.mediaAssetContent(id), options: options),
        MediaParts.file(input.bytes, input.filename, input.contentType),
      ),
    );
    return this;
  }

  /// Refused while an entry references it, unless forced.
  Future<void> delete([DeleteOptions options = const DeleteOptions()]) => _transport.empty(
    TransportRequest('DELETE', ApiPath.mediaAsset(id), query: {if (options.force) 'force': true}, options: options),
  );

  Future<MediaUsagePage> usages([
    MediaUsageQuery query = const MediaUsageQuery(),
    RequestOptions options = const RequestOptions(),
  ]) => MediaUsagePage.load(_transport, id, query, options);

  Future<MediaAsset> refresh([RequestOptions options = const RequestOptions()]) async {
    _record = MediaAssetRecord.fromJson(
      await _transport.json(TransportRequest('GET', ApiPath.mediaAsset(id), options: options)),
    );
    return this;
  }

  Future<MediaAsset> _patch(Map<String, Object?> body, RequestOptions options) async {
    _record = MediaAssetRecord.fromJson(
      await _transport.json(TransportRequest('PATCH', ApiPath.mediaAsset(id), body: body, options: options)),
    );
    return this;
  }
}
